test/abstract_client.js

```diff
@@ -780,6 +780,44 @@ module.exports = function (server, config) {
     it('should publish 10 QoS 2 and receive them only when `handleMessage` finishes', function (done) {
       testQosHandleMessage(2, done)
     })
+
+    it('should not send a `puback` if the execution of `handleMessage` fails', function (done) {
+      var client = connect()
+
+      var messageId = Math.floor(65535 * Math.random())
+      var payload = 'test'
+      var topic = 'test'
+      var qos = 1
+
+      var pubackReceived = false
+
+      client.handleMessage = function (packet, callback) {
+        callback(new Error('Error thrown by the application'))
+      }
```

Comment:
as we are erroring, should a PUBACK be sent anyway?

Reply:
IMHO, in an event where the handleMessage throws an error (could be due to the unavailability of the backend system that persists incoming messages, for example) we need to let the broker retry the messages at a later point. If we do send a PUBACK even amidst an error thrown by the application, does it not send a wrong indication to the broker that the message has successfully been processed by the application? WDYT?

Reply:
You are right, no PUBACK if handleMessage errors.